tests/testthat/test-compare.R

```diff
@@ -258,3 +258,73 @@ test_that("rewrite compare for gpu complains if indenting is bad", {
     transform_compare_odin_gpu(code),
     "Detected inconsistent indenting while reformatting compare function")
 })
+
+
+test_that("build compare with new interface", {
+  gen <- odin_dust({
+    initial(y) <- 0
+    update(y) <- y + rnorm(0, 1)
+    scale <- user(1)
+    observed <- data()
+    compare(observed) ~ normal(y, scale)
+  })
+
+  t <- seq(0, 50, by = 5)[-1]
+  d <- dust::dust_data(
+    data.frame(time = t,
+               observed = rnorm(length(t), 0, sqrt(t)),
+               another = 0L))
+
+  mod <- gen$new(list(), 0, 10)
+  expect_null(mod$compare_data())
+  mod$set_data(d)
+  expect_null(mod$compare_data())
+  y <- mod$run(t[[1]])
+  expect_equal(
+    mod$compare_data(),
+    dnorm(d[[1]][[2]]$observed, drop(y), 1, TRUE))
+})
+
+
+test_that("build compare with new interface for ode models", {
+  gen <- odin_dust({
+    # nolint start
+    deriv(N) <- r * N * (1 - N / K)
+    initial(N) <- N0
+    N0 <- user(1)
+    K <- user(100)
+    r <- user(0.1)
+    sd_noise <- user(0.5)
+    observed <- data()
+    compare(observed) ~ normal(N, sd_noise)
+    # nolint end
+  })
+
+  mod <- gen$new(list(), 0, 1)
+  t <- seq(0, 100, by = 5)[-1]
+  d <- dust::dust_data(
+    data.frame(time = t,
+               observed = runif(length(t), 0, 100)))
+  expect_null(mod$compare_data())
+  mod$set_data(d)
+  expect_null(mod$compare_data())
+  y <- mod$run(t[[1]])
+  expect_equal(
+    mod$compare_data(),
+    dnorm(d[[1]][[2]]$observed, drop(y), 0.5, TRUE))
```

Comment:
Note for self - the gaps in my maths/stats make it not quite intuitive why/what/how these two are equal - but they are, and it works and that's how the comparison succeeds.